### ISSUE TITLE
Bookshelves

### ISSUE THREAD

**Issue by nofreewill42** [pgcorpus/gutenberg#38]
Sadly the structure of bookshelves have changed on the site and hasn't been updated yet to match the old version.

http://www.gutenberg.org/wiki/Category:Bookshelf -> https://www.gutenberg.org/ebooks/bookshelf/
`Following a server upgrade in 2020, bookshelf listings are presented in the “flat” display below. Previously, there was a hierarchy, which has not yet been recreated.`

Books and their metadata have been downloaded, so it's just the bookshelf categories that are missing, not a big deal. Thank you for this repo!
Just wrote it down here so anybody coming across it can find it out easier.

**Comment by LJK150030** [comment]
That URL is only referenced in bookshelf.py on line 27

`    sp_args = ["wget",
               "--random-wait", "-r",
               "-A", "*Bookshelf*",
               "-p", "--no-parent",
               "-e", "robots=off",
               "-U", "mozilla",
               "http://www.gutenberg.org/wiki/Category:Bookshelf"
               ]
    subprocess.call(sp_args)`

In an attempt to reuse this code, I've altered that line with what @nofreewill42 suggested. However, I keep getting an error saying.

`mv: cannot stat ‘https://www.gutenberg.org/ebooks/bookshelf/index.html’: No such file or directory
Traceback (most recent call last):
  File "get_data.py", line 142, in <module>
    bookshelves_df = parse_bookshelves()
  File "/users/jklinkert/ondemand/data/sys/myjobs/projects/default/14/src/bookshelves.py", line 73, in parse_bookshelves
    PGids = list(set(np.concatenate(list(BS_dict.values()))))
  File "<__array_function__ internals>", line 6, in concatenate
ValueError: need at least one array to concatenate` 
  
Each dir in data has a dummy file, so I am unsure why this error is coming up.

Is this the only place to update to work with the new website structure, or am I missing something?

**Comment by alevas** [comment]
Hi!

I also ran into this issue and examined the `bookshelves.py` file a bit further. My problem is that I do not understand what is the end goal of the method `parse_bookshelves`.

I can now download every webpage `https://www.gutenberg.org/ebooks/bookshelf/XXX`, where XXX is any number corresponding to a valid bookshelf. Can anyone elaborate over the expected contents and use of `metadata/bookshelves.p`?

If I get this right, I can make a PR with the code changes that I had to do to make it work.
